llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -5833,6 +5834,80 @@ PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
   return DAG.getBitcast(VT, PRMT);
 }
 
+static SDValue
+PerformBUILD_VECTOROfTargetLoadCombine(SDNode *N,
+                                       TargetLowering::DAGCombinerInfo &DCI) {
+  // Match: BUILD_VECTOR of v4i8, where first two elements are from a
+  // NVPTXISD::LoadV2 or NVPTXISD::LDUV2 of i8, and the last two elements are
+  // zero constants. Replace with: zext the loaded i16 to i32, and return as a
+  // bitcast to v4i8.
```

Comment:
This appears to be an oddly specific pattern with two specific elements loaded, and two specific constants. Is that a particularly common pattern? Where does it come from, if so?

Reply:
> Is that a particularly common pattern?

This is obviously subjective, but I would not say it's a "common" pattern. 

> Where does it come from, if so?

This is an artifact of type legalization converting `vNi8` operations into `v4i8`. [Godbolt Link](https://godbolt.org/z/PcPv1G6r1)

What we really want to optimize is the series of `prmt`s that reconstruct the 16-bit load. This `BUILD_VECTOR` pattern is what generates the series of `prmt`s in all cases that I'm aware of. Thus, as an implementation decision, I wrote this peephole for this very specific `BUILD_VECTOR` case. 

(Previously, I wrote the code for this peephole by looking at the `prmt`s and it's a little bit more involved.)